### PR TITLE
Sorting and descriptions

### PR DESCRIPTION
**Pull request by gmn42**
Several improvements and changes, including sorting of tree and element lists and improving descriptions of elements, among others:

- Refactored project loading for additional speedup and added profiling routines
- Add nextPhraseID conversation reward type (requested by @NutAndor )
- Add friendly names for reward types, requirement types, and skills (using game descriptions)
- Sort comboboxes alphabetically, including GDE comboboxes.
- Sort main treeview (required adding mapping layer).  Note: Order of project folders changed due to sorting; 'Altered' is now above 'Created'.  Fixing this would require some klugery, but can be done if needed.  'AT Source' was changed to 'Game Source' to push it to the bottom of the list.
- implement keyboard focus traversal for text area in Editor (requested by Antison)
- Add keyboard focus border to Collapsible Panels
- Allow creating a project when directory already exists (treated as "import") (Requested by Johny & @NutAndor )
- Allow leaving project directory/data files in place when deleting project.

### CODE REVIEW

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 1]
```diff
@@ -314,7 +316,9 @@ static String getTreeNodeStateKey(ProjectTreeNode node) {
             GameDataElement gameDataElement = (GameDataElement) node;
             return "element:" + node.getClass().getName() + ":" + gameDataElement.id;
         }
-        return "node:" + node.getClass().getName() + ":" + node.getDesc();
+        String desc = node.getDesc();
+        if (desc != null && desc.startsWith("*")) desc = desc.replaceFirst("^\\*+", "");
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: I feel like regex is a bit overkill for this? we could just remove the first character since we already checked if it starts with a *. should be better readable.
Or at least add a comment documenting what the regex is doing. 
If going with the regex, remove the startsWith as the regex already has the ^ and then remove the + as I don't think that is needed right?

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 2]
```diff
@@ -509,33 +521,200 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
+         */
+        private java.util.Comparator<ProjectTreeNode> comparatorForParent(final ProjectTreeNode parent) {
+            return (a, b) -> {
+                String da = a == null ? "" : a.getDesc();
+                String db = b == null ? "" : b.getDesc();
+                if (da == null) da = "";
+                if (db == null) db = "";
+                // Ignore leading '*' markers (modified indicator) when sorting so saving state doesn't change order
+                String na = da.startsWith("*") ? da.replaceFirst("^\\*+", "") : da;
+                String nb = db.startsWith("*") ? db.replaceFirst("^\\*+", "") : db;
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: same thing there as in line 320

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 3]
```diff
@@ -509,33 +521,200 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
-                l.treeNodesChanged(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath(),
-                                                      new int[]{((ProjectTreeNode) node.getParentPath().getLastPathComponent()).getIndex((ProjectTreeNode) node.getLastPathComponent())},
-                                                      new Object[]{node.getLastPathComponent()}));
+            ProjectTreeNode parent = (ProjectTreeNode) node.getParentPath().getLastPathComponent();
+            ProjectTreeNode child = (ProjectTreeNode) node.getLastPathComponent();
+
+            List<ProjectTreeNode> list = sortedChildren(parent);
+            // find old index by identity
+            int oldIndex = -1;
+            for (int i = 0; i < list.size(); i++) {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:658]: is there a reason why we are searching manually here, while using binarySearch after removing and in insertNode?

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 4]
```diff
@@ -509,33 +521,200 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
+            int newIndex = bs < 0 ? -bs - 1 : bs;
+            list.add(newIndex, child);
+
+            // If position changed, fire removed + inserted; otherwise fire changed
+            if (oldIndex >= 0 && oldIndex != newIndex) {
+                for (TreeModelListener l : listeners) {
+                    l.treeNodesRemoved(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath(),
+                                                          new int[]{oldIndex},
+                                                          new Object[]{node.getLastPathComponent()}));
+                    l.treeNodesInserted(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath().getPath(),
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: any reason why the call to `getPath()` is here, while not calling it a few lines earlier?

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 5]
```diff
@@ -509,33 +521,200 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
+         * Handle removal of {@code node} from the tree model.
+         * <p>
+         * Attempts to remove the child from the cached sorted list using binary
+         * search; if found the list entry is removed and the removed event uses the
+         * sorted index. If the child is not found in the cache we fall back to the
+         * underlying parent's index for the TreeModelEvent.
+         *
+         * @param node the removed node's TreePath
+         */
         public void removeNode(TreePath node) {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:702]: I don't really like this whole method. It just seems like a very easy way to introduce errors.
For example, is it really the same if we pass the sortedIndex to the treeNodesRemoved event? This is the index in the sorted list, not necessarily the actual index right? And if it is the actual index, why do we need to sort it in the first place?

I feel like the whole sortedChildren logic is extremely easy to mess up, not even counting any background thread issues.

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java [hunk 6]
```diff
@@ -177,12 +178,15 @@ public void actionPerformed(ActionEvent e) {
...
+            // We pass callbacks to set/reset the busy flag when the actual creation occurs, to lock out the tree
+            new ProjectCreationWizard(() -> setWorkspaceBusy(true), () -> setWorkspaceBusy(false)).setVisible(true);
         }
 
     };
 
 
+    // TODO: Fix the memory leak here - it doesn't seem to usually free much of anything
     public ATCSAction closeProject = new ATCSAction("Close Project", "Closes the project, unloading all resources from memory") {
         public void actionPerformed(ActionEvent e) {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: don't we wanna check if the workspace is busy here?

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java [hunk 7]
```diff
@@ -196,28 +200,33 @@ public void selectionChanged(ProjectTreeNode selectedNode, TreePath[] selectedPa
...
         public void selectionChanged(ProjectTreeNode selectedNode, TreePath[] selectedPaths) {
-            setEnabled(selectedNode instanceof ClosedProject);
+            setEnabled(!workspaceBusy && selectedNode instanceof ClosedProject);
         }
 
     };
 
-    public ATCSAction deleteProject = new ATCSAction("Delete Project", "Deletes the project, and all created/altered data, from disk") {
+    public ATCSAction deleteProject = new ATCSAction("Delete Project", "Deletes the project from the workspace, optionally removing its folder from disk") {
         public void actionPerformed(ActionEvent e) {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: don't we wanna check if the workspace is busy here?

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java [hunk 8]
```diff
@@ -596,6 +605,21 @@ public void selectionChanged(ProjectTreeNode selectedNode, TreePath[] selectedPa
 
     };
 
+    /**
+     * worspaceBusy routines -
+     * Whether the workspace is currently busy (generally, load operation in progress)
+     */
+    private volatile boolean workspaceBusy = false;
+
+    private boolean isWorkspaceBusy() {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: is this ever used? I think it is always accessed directly

File: src/com/gpl/rpg/atcontentstudio/utils/Profiling.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+     *
+     * @param groupLabel label used for the summary line and optional per-item lines
+     * @param items items to process
+     * @param enabled whether the summary timing output should be emitted
+     * @param itemLabel function producing a readable label for each item
+     * @param action work to execute for each item
+     * @param <T> item type
+     * @return number of processed items
+     */
+    public static <T> int runEachParallel(String groupLabel, Collection<T> items, boolean enabled, Function<T, String> itemLabel, Consumer<T> action) {
```
Comment (OMGeeky) [inline src/com/gpl/rpg/atcontentstudio/utils/Profiling.java:100]: this is unused right?

**Review by OMGeeky — COMMENTED** [review]
First of all, thank you for all your work! It seems like you are really improving a lot of things here.

I like most of what you did, except maybe the sortedCache stuff in the ProjectsTree. That just seems like an extremely easy way to create issues for yourself with wrong indexes, out of bounds errors etc.
I did not fully understand all of it, since it's been a while since I took a look at ATCS and I don't really have too much time, but on first look it seems risky, and some AI code review tools also said there might be issues with thread safety and wrong event being called, because things not being in the right order, which even if not actual issues in our case, just confirms my intuition, that this is easy to loose track of when what is happening.

Besides that, I added some smaller notes, which should all be pretty straight forward.

**Comment by OMGeeky** [comment]
Just for completeness, here is a code review report an AI created for me, which might be interesting:


---

## Critical Logic Issues & Bugs

### 1. Project Corruption & NPE on Legacy Project Load
* **File:** [Project.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/model/Project.java)
* **Location:** `fromFolder(Workspace w, File projRoot)` method (Lines 245–263)
* **Bug Description:** The PR removes the call to `p.refreshTransients(w);` when loading an obsolete binary `.project` file. 
* **Impact:** 
  - Standard Java deserialization does not restore transient fields. Therefore, `parent` (Workspace parent), `alteredContent`, `createdContent`, `bookmarks`, `v`, and `projectElementListeners` will all remain `null`.
  - `p.baseContent.parent` will also remain `null` since the `refreshTransients(this)` call on `baseContent` was removed.
  - Calling `p.save()` writes this partially uninitialized state as JSON to `.project.json`, permanently corrupting the project configuration. Subsequent operations on the returned project will immediately trigger `NullPointerException`s.

### 2. Missing Dialogue Linking for `setNextPhraseID` Reward Type
* **File:** [Dialogue.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/model/gamedata/Dialogue.java)
* **Location:** `link()` method (Lines 281–343)
* **Bug Description:** The new reward type `setNextPhraseID` is parsed from JSON and added to the UI, but it was not added to the `switch (reward.type)` statement in the `link()` method.
* **Impact:** 
  - `reward.reward_obj` is never resolved to the target dialogue object when loading projects, leaving it `null`.
  - Because `reward.reward_obj` is null, the target dialogue does not receive a backlink to the source dialogue.
  - This breaks the dialogue graph visualization, target-reference checking, and refactoring/renaming propagation for dialogues linked via the `setNextPhraseID` reward.

### 3. NullPointerException on Load Failure in `Workspace.java`
* **File:** [Workspace.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/model/Workspace.java)
* **Location:** `openProject(final ClosedProject cp, final Runnable onFinished)` (Lines 314–336)
* **Bug Description:** `Project.fromFolder(...)` can return `null` if the project root folder does not contain a `.project` or `.project.json` file. The code immediately calls `p.open();` on the returned instance without performing a null check.
* **Impact:** If the project directory is corrupted or empty, a `NullPointerException` will be thrown inside the background thread of `WorkerDialog`.

### 3.1. ProjectsTreeModel.removeNode Binary Search Failure & UI Collapse Bug
* **File:** [ProjectsTree.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java)
* **Location:** `removeNode(TreePath node)` method (Lines 683–710)
* **Bug Description:** 
  - `removeNode` uses `Collections.binarySearch(list, child, cmp)` to find the index of the child to remove in the sorted cache.
  - However, in ATCS, when a project is deleted, it is removed from the parent node's list in the model *before* `removeNode` is called.
  - As a result, when the comparator executes during the binary search, `parent.getIndex(child)` returns `-1`. This causes the deleted child node to compare inconsistently with the other children, violating the binary search prerequisite of a sorted list.
  - The binary search fails and returns a negative value. The fallback code then calls `parent.getIndex(child)`, which also returns `-1`.
* **Impact:** Because the determined index is `-1` (< 0), the model triggers a full `treeStructureChanged()` event instead of a clean `treeNodesRemoved()` event. This causes Swing's `JTree` to collapse and reload the parent's entire subtree, forcing the user to manually re-expand the projects tree branch every time a project is deleted.
* **Fix:** Look up the child's index in the cached list by identity (a simple loop/indexOf) instead of using binary search.

### 3.2. ProjectsTreeModel getChild/getChildCount TreeModel Contract Violation (Latent Bug)
* **File:** [ProjectsTree.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java)
* **Location:** `getChild` and `getChildCount` methods (Lines 502–512)
* **Bug Description:** 
  - `getChildCount(parent)` returns `parent.getChildCount()`, which is the total count of children in the underlying model.
  - `getChild(parent, index)` returns the child from the lazily built `sortedChildren(parent)` list, which defensively filters children to only include instances of `ProjectTreeNode` via `t instanceof ProjectTreeNode`.
  - In the current codebase, all tree nodes implement `ProjectTreeNode`, meaning the list sizes match and no active crashes occur. However, if any child node that does not implement `ProjectTreeNode` is introduced in the future, the sorted children list will become shorter than the count returned by `getChildCount(parent)`.
* **Impact:** This is a future maintenance hazard. If non-`ProjectTreeNode` children are added, `getChild(...)` will return `null` when the JTree queries indices past the end of the filtered list. Returning `null` for a valid index `0 <= index < getChildCount()` violates the `TreeModel` contract, causing rendering glitches and layout cache failures in JTree.
* **Fix:** Change `getChildCount(parent)` to return the size of the sorted list (`sortedChildren(parent).size()`) instead of the raw underlying count, or remove the defensive filtering if all children are guaranteed to implement `ProjectTreeNode`.

---

## UI Rendering & Styling Issues

### 4. DashedBorder Opacity Glitch in Titled Panel
* **File:** [CollapsiblePanel.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/CollapsiblePanel.java)
* **Location:** Nested `DashedBorder` record (Lines 17–42)
* **Bug Description:** The custom `DashedBorder` implements the `Border` interface and overrides `isBorderOpaque()` to return `true`:
  ```java
  private record DashedBorder(Color color, int thickness, float[] dash) implements Border {
      @Override
      public boolean isBorderOpaque() {
          return true;
      }
  ```
* **Impact:** Under Swing's rendering model, returning `true` for `isBorderOpaque()` promises that the border paints every pixel in its insets. Since `DashedBorder` only draws a thin dashed rectangle (`g2.drawRect(...)`), it leaves most pixels transparent. This causes Swing to skip painting the background underneath the border, resulting in visual rendering artifacts, garbage pixels, and trailing graphics whenever the panel updates or receives focus.

---

## Concurrency & Thread-Safety Violations

### 5. Multi-Threaded Cache Mutation & Iteration Races
* **File:** [ProjectsTree.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java)
* **Location:** `ProjectsTreeModel` class (Lines 489–620)
* **Bug Description:** The new sorting mechanism uses `WeakHashMap` and `Collections.synchronizedList` to cache sorted lists of tree nodes. However:
  - Methods like `changeNode` perform non-atomic checks and iterations (loops) over `list` without external synchronization block guards.
  - If project data modifications occur on background worker threads (like `Workspace.createProject` or `Workspace.openProject`), the cache list can be structurally modified concurrently with EDT reads (e.g. for JTree rendering).
* **Impact:** 
  - Iteration loops can throw `IndexOutOfBoundsException` or `ConcurrentModificationException` if the list size changes.
  - JTree calls `getChild` and `getIndexOfChild` which read from the list; reading a mutating list can return inconsistent elements, leading to tree rendering mismatches.
  - During sorting, the comparator uses `parent.getIndex(TreeNode)` as a tie-breaker. If the underlying children lists are mutated concurrently, the comparator can return inconsistent values, throwing `IllegalArgumentException: Comparison method violates its general contract` (TimSort contract violation).

### 6. Background Thread Swing Mutation (EDT Violations)
* **Files:** [ProjectsTree.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java), [WorkspaceActions.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java)
* **Location:** `ProjectsTreeModel.insertNode / removeNode / changeNode` and `WorkspaceActions.setWorkspaceBusy`
* **Bug Description:** 
  - `Workspace.openProject` and `Workspace.createProject` invoke tree structure notifications (`p.notifyCreated()`, which bubbles up to `ProjectsTreeModel.insertNode`) inside the background task of a `WorkerDialog`.
  - `insertNode`, `removeNode`, and `changeNode` fire Swing tree model events (`treeNodesInserted`, etc.) directly on whatever thread calls them.
  - Similarly, background thread callbacks (like `onFinished`) invoke `setWorkspaceBusy()`, which directly updates Swing action states via `selectionChanged()` on the calling thread.
* **Impact:** Mutating Swing components or firing model events from a non-EDT thread violates Swing's single-threaded model, leading to deadlocks, UI hangs, and random exceptions in Swing's UI thread.

---

## Clean Code & Compilation Cleanup

### 7. Unused Imports
* **Files & Imports:**
  - `WorkspaceActions.java` (Line 4): `import com.gpl.rpg.atcontentstudio.Notification;`
  - `DialogueEditor.java` (Line 22): `import java.util.Comparator;`
  - `GameSource.java` (Line 5): `import com.gpl.rpg.atcontentstudio.model.gamedata.GameDataCategory;`
* **Impact:** These are unused imports. While they compiled under Java 17, they should be cleaned up to avoid IDE warnings and potential compilation failures in environments configured with strict `-Werror` (treat warnings as errors) compiler options.

### 8. Typo in Javadoc
* **File:** [ProjectsTree.java](file:///home/frb/git/personal/AT/ATCS/src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java)
* **Location:** Line 498
* **Correction:** `@param parent - The note...` should be `@param parent - The node...`.

File: src/com/gpl/rpg/atcontentstudio/utils/Profiling.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    public static void run(String label, boolean enabled, Runnable action) {
+        long start = enabled ? System.nanoTime() : 0L;
+        increaseIndent();
+        action.run();
+        if (enabled) {
+            printf("%s took %d ms", label, elapsedMillis(start));
+        }
+        decreaseIndent();
+    }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/utils/Profiling.java:35]: applied

File: src/com/gpl/rpg/atcontentstudio/ui/gamedataeditors/DialogueEditor.java [hunk 11]
```diff
@@ -1,1417 +1,1478 @@
...
+
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.awt.event.ItemEvent;
+import java.awt.event.ItemListener;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/gamedataeditors/DialogueEditor.java]: applied

File: src/com/gpl/rpg/atcontentstudio/ui/Editor.java [hunk 12]
```diff
@@ -270,6 +266,31 @@ public static JTextArea addTextArea(JPanel pane, String label, String initialVal
         tfArea.setLineWrap(true);
         tfArea.setWrapStyleWord(true);
 
+        // Ensure Tab moves focus instead of inserting a tab character
+        tfArea.setFocusTraversalKeysEnabled(true);
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/Editor.java]: Suggestion is technically wrong (it was working anyway), but we probably still should suppress built-in handling.  Applied and verified working.

File: src/com/gpl/rpg/atcontentstudio/model/Project.java [hunk 13]
```diff
@@ -240,38 +259,30 @@ public static Project fromFolder(Workspace w, File projRoot) {
             }
             p.save();
         }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/model/Project.java]: refactored to convert legacy to .json then load using main path.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 14]
```diff
@@ -489,9 +493,17 @@ public Object getRoot() {
             return Workspace.activeWorkspace;
         }
 
+        /**
+         * Return child according to a sorted view of the parent's children.
+         * @param parent - The note whose children should be returned
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: fixed

File: src/com/gpl/rpg/atcontentstudio/model/Project.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/model/Project.java]: Also, added a regression test for a round-trip save-and-load using the legacy savefile format

File: src/com/gpl/rpg/atcontentstudio/model/GameSource.java [hunk 15]
```diff
@@ -2,6 +2,7 @@
 
 import com.gpl.rpg.atcontentstudio.io.JsonSerializable;
 import com.gpl.rpg.atcontentstudio.model.Project.ResourceSet;
+import com.gpl.rpg.atcontentstudio.model.gamedata.GameDataCategory;
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/model/GameSource.java]: done

File: src/com/gpl/rpg/atcontentstudio/utils/Profiling.java [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        try {
+            action.run();
+        }finally {
+            if(enabled) {
+                printf("%s took %d ms", label, elapsedMillis(start));
+            }
+            decreaseIndent();
+        }
+        decreaseIndent();
+    }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/utils/Profiling.java:35]: applied for real  this time

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java [hunk 17]
```diff
@@ -596,6 +605,21 @@ public void selectionChanged(ProjectTreeNode selectedNode, TreePath[] selectedPa
...
+    private volatile boolean workspaceBusy = false;
+
+    private boolean isWorkspaceBusy() {
+        return workspaceBusy;
+    }
+
+    public void setWorkspaceBusy(boolean busy) {
+        workspaceBusy = busy;
+        selectionChanged(selectedNode, selectedPaths);
+    }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: Applied.  Needed since selectionChanged() enables/disables menu items, so needs to run on EDT.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java (hunk 1 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: Maybe a little heavy, but checking the * explicitly first keeps the regex out of the hot path.  Not really pretty though, it was an autocomplete I didn't bother to make cleaner.  But you're right, it should only match the first character, so the + was a bug.  I've changed it to just use `desc.substring(1)` (and also fixed the comparator).

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java (hunk 2 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: fixed, using `substring(1)` now.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java (hunk 3 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:658]: Because it's already in the list, and the node may have been renamed.  A binary search would find where the node needs to go now, not the current location we need to remove it from.  Since we don't have the old name, we need to crawl the list looking for the object identity itself.   This also avoids ambiguity if duplicate names are ever allowed for some reason.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java (hunk 4 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java]: No particular reason, I think I must have cleaned up one and not the other.  They end up doing the same thing under the hood (they just trigger different overloads on the same constructor), but I cleaned it up for consistency here and also in removeNode().

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java (hunk 5 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:702]: Good catch on the index, it was a "best effort" approach that shouldn't have been there.  I've removed that and will just fire `treeStructureChanged` and print a stack trace if the node can't be found.

The cached sort layer admittedly is somewhat fragile; we need to make sure that it catches all update events to keep the trees in sync, and we need to make sure it stays single threaded - really called through the EDT only,but that's already the case for a swing object.  There are some issues with that in the existing code doing swing stuff off-EDT, that's actually how I got started down the road with the workspaceBusy() stuff that I'm still not happy with and want to look at more carefully.

Before going with this approach, I considered the alternative of sorting the underlying data, but that would have meant altering each of the 6 collection types to apply sorting on adds, changes, deletes, etc.  The cache avoids tightly coupling the the display and storage layers and keeps everything in one place.

So far it's been reliable in my testing, and I'll add some unit tests to try to flush out any edge cases that might be in there.

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java (hunk 6 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: yeah, we should for completeness, done.  I'm not really happy with the whole workspaceBusy mechanism and didn't fully implement it.  I want to figure out a better way to guard access to the tree without disabling the menu items, but this should prevent the crashing for now.

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java (hunk 7 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: also done

File: src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java (hunk 8 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/WorkspaceActions.java]: cleaned up.

File: src/com/gpl/rpg/atcontentstudio/utils/Profiling.java (hunk 9 above)
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/utils/Profiling.java:100]: Yes, for now.  I'm hoping to use it to parallelize the linking during project load, but I haven't got there yet.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 18]
```diff
@@ -489,14 +493,22 @@ public Object getRoot() {
...
+            if (!(parent instanceof ProjectTreeNode)) return null;
+            List<ProjectTreeNode> children = sortedChildren((ProjectTreeNode) parent);
+            return (index >= 0 && index < children.size()) ? children.get(index) : null;
         }
 
         @Override
         public int getChildCount(Object parent) {
-            return ((ProjectTreeNode) parent).getChildCount();
+            return (sortedChildren((ProjectTreeNode) parent)).size();
         }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:515]: not an issue in this codebase; we only use ProjectTreeNodes.

File: src/com/gpl/rpg/atcontentstudio/ui/Editor.java [hunk 19]
```diff
@@ -270,6 +266,31 @@ public static JTextArea addTextArea(JPanel pane, String label, String initialVal
...
+        // Ensure Tab moves focus instead of inserting a tab character
+        tfArea.setFocusTraversalKeysEnabled(false);
+        InputMap im = tfArea.getInputMap(JComponent.WHEN_FOCUSED);
+        // Bind Tab and Shift+Tab to focus traversal actions (transferFocus/transferFocusBackward)
+        im.put(KeyStroke.getKeyStroke(KeyEvent.VK_TAB, 0), "focusNextComponent");
+        im.put(KeyStroke.getKeyStroke(KeyEvent.VK_TAB, KeyEvent.SHIFT_DOWN_MASK), "focusPreviousComponent");
+        tfArea.getActionMap().put("focusNextComponent", new AbstractAction() {
+            @Override
+            public void actionPerformed(ActionEvent e) { Editor.focusTraversalRequested = true; tfArea.transferFocus(); }
+        });
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/Editor.java:276]: removed logic requiring this flag, wasn't really a good idea anyway.

File: src/com/gpl/rpg/atcontentstudio/ui/Editor.java [hunk 20]
```diff
@@ -270,6 +266,31 @@ public static JTextArea addTextArea(JPanel pane, String label, String initialVal
...
+        tfArea.getActionMap().put("focusPreviousComponent", new AbstractAction() {
+            @Override
+            public void actionPerformed(ActionEvent e) { Editor.focusTraversalRequested = true; tfArea.transferFocusBackward(); }
+        });
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/Editor.java]: removed logic requiring this flag, wasn't really a good idea anyway.

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectCreationWizard.java [hunk 21]
```diff
@@ -259,26 +261,38 @@ protected void updateOkButtonEnablement() {
             this.okButton.setEnabled(false);
             return;
         }
+        String projectName = projectNameField.getText();
+        if (!projectName.equals(projectName.trim())) {
+            errorLabel.setText("<html><font color=\"#FF0000\">Project names cannot start or end with spaces.</font></html>");
+            this.okButton.setEnabled(false);
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectCreationWizard.java]: good call.  Added better project name validation,

File: src/com/gpl/rpg/atcontentstudio/ui/gamedataeditors/DialogueEditor.java [hunk 22]
```diff
@@ -1216,7 +1229,11 @@ public void valueChanged(JComponent source, Object value) {
                     return;
                 }
             } else if (source == messageField) {
-                dialogue.message = (String) value;
+                if(((String) value).isEmpty()) {
+                    dialogue.message = null;
+                } else {
+                    dialogue.message = (String) value;
+                }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/gamedataeditors/DialogueEditor.java]: fixed

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 23]
```diff
@@ -509,33 +524,218 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
+                                                           new Object[]{node.getLastPathComponent()}));
+                }
+            } else {
+                int idx = newIndex >= 0 ? newIndex : (oldIndex >= 0 ? oldIndex : parent.getIndex((TreeNode) child));
+                for (TreeModelListener l : listeners) {
+                    l.treeNodesChanged(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath(),
+                                                          new int[]{idx},
+                                                          new Object[]{node.getLastPathComponent()}));
+                }
             }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:690]: fixed

File: src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java [hunk 24]
```diff
@@ -509,33 +524,218 @@ public void valueForPathChanged(TreePath path, Object newValue) {
...
+                sortedIndex = bs;
+            }
+
             for (TreeModelListener l : listeners) {
-                l.treeNodesInserted(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath().getPath(),
-                                                       new int[]{((ProjectTreeNode) node.getParentPath().getLastPathComponent()).getIndex((ProjectTreeNode) node.getLastPathComponent())},
+                l.treeNodesInserted(new TreeModelEvent(node.getLastPathComponent(), node.getParentPath(),
+                                                       new int[]{sortedIndex},
                                                        new Object[]{node.getLastPathComponent()}));
             }
```
[the remark this answers is not in the document]
Comment (gmn42) [reply src/com/gpl/rpg/atcontentstudio/ui/ProjectsTree.java:637]: fixed
